refactor: Personal Data Deletion Request

gavindsouza opened the pull request:
# Changes

* Single trigger for Data Deletion workflow
* Disable user before data deletion process starts
* Standardize anonymization Data
* Redact field data completely defined via redact_fields
* Partial field value redaction (replace Full Name and User email ID from specified DocTypes)
* New options: `partial`, `rename`, `strict`
* Renamed options:  ~`personal_fields`~ => `redact_fields`, ~`match_field`~ => `filter_by`
* Removed option: `applies_to_website_user`

Docs: https://github.com/frappe/frappe_docs/pull/102

gavindsouza committed:
style: Re-fromatted functions via Black-ish

gavindsouza committed:
fix: Redact field data completely defined via redact_fields

* Replace data in the fields mentioned in the hook, user_data_fields'
redact_fields option
* The value is decided by utilizing the anonymization_value_map defined
by the __set_anonymization_data method
* Rename User doc with self.anon value and mark as "Deleted"

gavindsouza committed:
fix: Standardize anonymization Data

* Set anonymization matrix for instance of Personal Data Deletion Request
* During the initialization of the object, privacy docs are segregated
into full match and partial data redaction types
* Context for execution of _anonymize_data is set via
__set_anonymization_data by default unless set_data is False

gavindsouza committed:
fix: Single trigger for Data Deletion workflow

* Sytem Manager authorization for triggering data deletion
* Validate form verififcation state
* Disable user before data deletion process starts

gavindsouza committed:
fix: Use standard frappe configuration for Email validation

happy side-effect: JSON shrinks with new schema

gavindsouza committed:
refactor(hooks): Personal Data Deletion Request

Changes:
* Renamed hook from `user_privacy_documents` to `user_data_fields`
* Renamed ref field from `match_field` to `filter_by`
* Renamed ref field from `personal_fields` to `redact_fields`
* Removed ref field `applies_to_website_user` from hook options -- let's
treat all requests equally and redact data from all privacy docs
* Added boilerplate hook for new apps

gavindsouza committed:
refactor: Split functionality into re-usable methods

gavindsouza committed:
feat: Partial field value redaction

Finds and replaces Full Name and User email ID from specified DocTypes

Changes:
* Option "partial" added in user_data_fields
* If "redact_fields" aren't speciifed, "partial" mode is assumed
* If "rename" is set, the respective docs are renamed with self.anon
* "strict" if unset, is assumed to be False. In this case, a non conditional
query is used to delete data. If "strict" is True, Personal Data
Deletion Request will obey "filter_by" field value if defined else "owner" is used

gavindsouza committed:
style: Prettier JS styling

gavindsouza committed:
fix: Set anonymization dict during init

Alchez commented on frappe/utils/boilerplate.py:
```diff
@@ -278,6 +278,31 @@ def make_boilerplate(dest, app_name):
 # auto_cancel_exempted_doctypes = ["Auto Repeat"]
 
+
+# User Data Protection
```
Should this happen via Doctype configurations instead? That might make it easier for users to adapt to policy changes as well.

gavindsouza replied:
Yeah. The thought is to have Data Protection Settings at a docfield level. But this may need to have more iterations to get there. For now, I'm treating this refactor as the first step towards management via configurations.

gavindsouza committed:
test: Personal Data Deletion Request

gavindsouza committed:
fix: Update hooks for user privacy documents

gavindsouza committed:
fix: Use site domain for deleted user name

* Use strict option for Comment

gavindsouza committed:
fix: Autoname documents with valid email IDs

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-personal-data-deletion

gavindsouza committed:
test: Personal Data Deletion Request

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-personal-data-deletion

surajshetty3416 commented:
Please re-open once this is ready for review.

gavindsouza committed:
ci: Update Postgres version to development server

gavindsouza committed:
ci: Install and run specific version of PostgreSQL

ref: https://docs.travis-ci.com/user/database-setup/#using-a-different-postgresql-version

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-personal-data-deletion

gavindsouza committed:
chore: Bump Postgres to 12.4

gavindsouza committed:
fix: Replace email and name data in single assignment

gavindsouza committed:
fix: Delete user after test is run

gavindsouza committed:
refactor: Re-order methods for better readability

surajshetty3416 committed:
Merge branch 'develop' into refactor-personal-data-deletion

gavindsouza committed:
chore: Boolean over Int for user_data_fields

gavindsouza committed:
Merge branch 'refactor-personal-data-deletion' of github.com:gavindsouza/frappe into refactor-personal-data-deletion

gavindsouza committed:
ci: Bump mariaDB version to 10.4

10.5 is latest stable, but this feature was developed on 10.4; so
leaving this here.

surajshetty3416 committed:
Merge branch 'develop' into refactor-personal-data-deletion

gavindsouza committed:
Revert "ci: Bump mariaDB version to 10.4"

This reverts commit c9d0aae791ff96c4487ecbebdc51eab697268b82.

gavindsouza committed:
Merge branch 'refactor-personal-data-deletion' of github.com:gavindsouza/frappe into refactor-personal-data-deletion

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-personal-data-deletion
